src/common/perf_counters_cache_key.cc

```diff
@@ -0,0 +1,214 @@
+#include <algorithm>
+#include <iterator>
+#include <numeric>
+#include <optional>
+#include "common/perf_counters_cache_key.h"
+
+namespace ceph::perf_counters::detail {
+
+// use a null character to delimit strings
+constexpr char delimiter = ':';
+
+
+// write a delimited string to the output
+auto write(std::string_view str, std::output_iterator<char> auto out)
+{
+  out = std::copy(str.begin(), str.end(), out);
+  *(out++) = delimiter;
+  return out;
+}
+
+// return the encoded size of a label
+inline std::size_t label_size(const label_pair& l)
+{
+  return l.first.size() + sizeof(delimiter)
+      + l.second.size() + sizeof(delimiter);
+}
+
+// a forward iterator over label_pairs encoded on a flat buffer
+class label_iterator {
+  using base_iterator = const char*;
+
+  struct iterator_state {
+    base_iterator pos; // end of current label
+    base_iterator end; // end of buffer
+    label_pair label; // current label
+
+    auto operator<=>(const iterator_state& rhs) const = default;
+  };
+  // an empty state represents a past-the-end iterator
+  std::optional<iterator_state> state;
+
+  // find the next two delimiters and construct the label string views
+  static void advance(std::optional<iterator_state>& s)
+  {
+    auto d = std::find(s->pos, s->end, delimiter);
+    if (d == s->end) { // no delimiter for label key
+      s = std::nullopt;
+      return;
+    }
+    s->label.first = std::string_view{s->pos, d};
+    s->pos = std::next(d);
+
+    d = std::find(s->pos, s->end, delimiter);
+    if (d == s->end) { // no delimiter for label name
+      s = std::nullopt;
+      return;
+    }
+    s->label.second = std::string_view{s->pos, d};
+    s->pos = std::next(d);
+  }
+
+  // try to parse the first label pair
+  static auto make_state(base_iterator begin, base_iterator end)
+  {
+    std::optional state = iterator_state{begin, end};
+    advance(state);
+    return state;
+  }
+
+ public:
+  using difference_type = std::ptrdiff_t;
+  using value_type = label_pair;
+  using pointer = const value_type*;
+  using reference = const value_type&;
+
+  label_iterator() = default;
+  label_iterator(base_iterator begin, base_iterator end)
+      : state(make_state(begin, end)) {
+    //static_assert(std::forward_iterator<label_iterator>);
```

Comment:
if these static assertions are failing, we should fix the types so they conform to their concepts